webui/react/src/pages/ExperimentDetails/TrialsComparisonModal.tsx

```diff
@@ -111,11 +124,15 @@ export const TrialsComparisonTable: React.FC<TableProps> = ({
     setTrialsDetails(trials);
   }, [trials]);
 
+  useEffect(() => {
+    if (runs === undefined) return;
+    setTrialsDetails(runs);
+  }, [runs]);
```

Comment:
this is inherited, but given `trialIds`, `trials` and `runs` are mutually exclusive, we can simplify the component by maintaining the `trialIds` request data separate from the props in `trials` and `runs`.

firstly, we rewrite the fetching logic to use `useAsync`:
```tsx
  // the loadable has the flat run type here to make the getOrElse later on succeed.
  const trialDetailsRequest = useAsync<TrialDetails[] | FlatRun[]>(async (canceler) => {
    if (trialIds === undefined) return NotLoaded;
    return (await Promise.all(trialIds.map(async (trialId: number) => {
      try {
        return await getTrialDetails({ id: trialId }, { signal: canceler.signal });
      } catch (e) {
        handleError(e);
        return undefined;
      }
    })).filter((r): r is Exclude<typeof r, undefined> => !!r);
  });
```
then, we can `getOrElse` that and centralize the fallback logic:
```tsx
  const trialDetails = useMemo(() => {
    return trialDetailsRequest.getOrElse(trials ?? runs ?? [])
  }, [trialDetailsRequest, trials, runs])
```

this has the upshot of saving renders for the `trials` and `runs` cases as well.

Reply:
Nice!